Add filtering option to `rustc_on_unimplemented` and reword `Iterator` E0277 errors

estebank opened the pull request:
- Add more targetting filters for arrays to `rustc_on_unimplemented` (Fix #53766)
 - Detect one element array of `Range` type, which is potentially a typo:
   `for _ in [0..10] {}` where iterating between `0` and `10` was intended.
   (Fix #23141)
 - Suggest `.bytes()` and `.chars()` for `String`.
 - Suggest borrowing or `.iter()` on arrays (Fix #36391)
 - Suggest using range literal when iterating on integers (Fix #34353)
 - Do not suggest `.iter()` by default (Fix #50773, fix #46806)
 - Add regression test (Fix #22872)

estebank committed:
Add more targetting filters for arrays to rustc_on_unimplemented

estebank committed:
Reword `rustc_on_unimplemented` errors for `Iterator`

 - Detect one element array of `Range` type, which is potentially a typo:
   `for _ in [0..10] {}` where iterating between `0` and `10` was intended.
   (#23141)
 - Suggest `.bytes()` and `.chars()` for `String`.
 - Suggest borrowing or `.iter()` on arrays (#36391)
 - Suggest using range literal when iterating on integers (#34353)
 - Do not suggest `.iter()` by default (#50773, #46806)

estebank committed:
Add regression test (#22872)

estebank committed:
Add tests for handled cases

rust-highfive commented:
r? @varkor

(rust_highfive has picked a reviewer for you, use r? to override)

estebank commented:
CC @csmoe I think this covers all that #52178 did

estebank committed:
fix tidy

varkor commented on src/libcore/iter/iterator.rs:
```diff
@@ -30,11 +30,32 @@ fn _assert_is_object_safe(_: &dyn Iterator<Item=()>) {}
+    on(
+        _Self="[std::ops::Range<Idx>; 1]",
+        label="if you meant to iterate between two values, remove the square brackets",
+        note="`[start..end]` is an array of one `Range`, you might have meant to have a `Range`: \
```
Semicolon is slightly more appropriate, I think:
```
`[start..end]` is an array of one `Range`; you might have meant to have a `Range`
```

varkor commented on src/libcore/iter/iterator.rs:
```diff
@@ -30,11 +30,32 @@ fn _assert_is_object_safe(_: &dyn Iterator<Item=()>) {}
 #[stable(feature = "rust1", since = "1.0.0")]
 #[rustc_on_unimplemented(
+    on(
+        _Self="[std::ops::Range<Idx>; 1]",
```
We should add these suggestions for the other types of range as well.

varkor reviewed (changes requested):
These diagnostic changes are a big improvement!

It'd be nice if we could get rid of the false positives (e.g. the "if you want to iterate between `0` until a value `end`" message), but it's hard to tell how often they'd come up in practice, so if that's awkward to support, I'd probably be okay with leaving it how it is: I think it probably helps more than it confuses.

estebank commented on src/libcore/iter/iterator.rs:
```diff
@@ -30,11 +30,32 @@ fn _assert_is_object_safe(_: &dyn Iterator<Item=()>) {}
+    on(
+        _Self="{integral}",
+        note="if you want to iterate between `0` until a value `end`, use the range syntax: \
+              `0..end`"
```
It's a note so it might be ok for it to be a bit long.

estebank commented on src/test/ui/issues/issue-50480.stderr:
```diff
@@ -4,13 +4,14 @@ error[E0412]: cannot find type `NotDefined` in this scope
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^ `i32` is not an iterator
    |
    = help: the trait `std::iter::Iterator` is not implemented for `i32`
+   = note: if you want to iterate between `0` until a value `end`, use the range syntax: `0..end`
```
That was my thinking.

estebank replied:
This would also be fixed by marking all the spans part of a for loop as being part of a desugaring context, which we can filter on in `rustc_on_unimplemented` and would improve the situation for many of this diagnostics that only apply when using a binding and not in type signatures.

varkor replied:
Ah, that would be good!

estebank commented on src/test/ui/iterators/array-of-ranges.stderr:
```diff
@@ -0,0 +1,43 @@
+  --> $DIR/array-of-ranges.rs:7:14
+   |
+LL |     for _ in array_of_range {}
+   |              ^^^^^^^^^^^^^^ if you meant to iterate between two values, remove the square brackets
```
We would have to change `rustc_on_unimplemented`. I've tried to "poison" the span for the iterator with a "desugaring context", but was unable to get it working. It should be possible to add it, but given that the likelihood of this case in particular happening is so slim (you have to 1. create an array of ranges, already uncommon and 2. try to iterate over the array), I felt confident leaving it as is (at least for now).

varkor replied:
That's true; it could always be a follow-up.

estebank committed:
review comments

- reword messages
- apply custom comments to all types of ranges
- fix indentation

varkor commented on src/libcore/iter/iterator.rs:
```diff
@@ -30,11 +30,56 @@ fn _assert_is_object_safe(_: &dyn Iterator<Item=()>) {}
+    ),
+    on(
+        _Self="[std::ops::RangeTo<Idx>; 1]",
+        label="if you meant to iterate until a value, remove the square brackets",
```
You can't iterate over a `RangeTo`, so I'm not sure this one is necessary.

estebank replied:
You're absolutely right, I just never had thought about it enough to look it up. I will still add a note to it in order to let users know that.

varkor commented on src/libcore/iter/iterator.rs:
```diff
@@ -30,11 +30,56 @@ fn _assert_is_object_safe(_: &dyn Iterator<Item=()>) {}
+    ),
+    on(
+        _Self="[std::ops::RangeToInclusive<Idx>; 1]",
+        label="if you meant to iterate until a value, remove the square brackets",
```
Same with `RangeToInclusive`.

varkor commented:
r=me after the last couple of comments, if Travis passes.

estebank committed:
Reword Range*/[Range*]: Iterator E0277 messages

estebank commented:
@bors r=varkor

bors commented:
:pushpin: Commit 5beeb535e1cbe0e93d623f674a33c57ed885c812 has been approved by `varkor`

<!-- @bors r=varkor 5beeb535e1cbe0e93d623f674a33c57ed885c812 -->

bors commented:
:hourglass: Testing commit 5beeb535e1cbe0e93d623f674a33c57ed885c812 with merge 73008786a90f5da800834e1540ca6521d8b7c7ba...

estebank committed:
Change Scalar to numeric cast

estebank commented:
@bors try

bors commented:
:hourglass: Trying commit c2478dfa477856a9bf7f3d4b617d781fef11b9d6 with merge 7959c3e7cd04fabd0503f810b9e329cb26fa5e35...

estebank commented:
For the record: the failing test was `iterators/array-of-ranges.rs` under `dist-i586-gnu-i586-i686-musl` due to an assertion in `Scalar::to_u64`:

```
[00:48:15] thread 'main' panicked at 'assertion failed: `(left == right)`
[00:48:15]   left: `8`,
[00:48:15]  right: `4`', librustc/mir/interpret/value.rs:242:17
[00:48:15] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:48:15] 
[00:48:15] error: internal compiler error: unexpected panic
```

estebank commented:
@bors try

bors commented:
:hourglass: Trying commit f5cc6b5cf8574c49148079320ce8d87432009448 with merge 2b66d93a47a16af1655e121b9a511f6ca1e06ebc...

bors commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang/rust/builds/441782992?utm_source=github_status&utm_medium=notification)
State: approved= try=True

bors commented:
:pushpin: Commit f5cc6b5cf8574c49148079320ce8d87432009448 has been approved by `varkor`

<!-- @bors r=varkor f5cc6b5cf8574c49148079320ce8d87432009448 -->

bors commented:
:hourglass: Testing commit f5cc6b5cf8574c49148079320ce8d87432009448 with merge d502144fd36634a938d73bf1893580d420194443...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/441956354?utm_source=github_status&utm_medium=notification)

rust-highfive commented:
The job `dist-i586-gnu-i586-i686-musl` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/441956372) ([raw log](https://api.travis-ci.org/v3/job/441956372/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Downloading https://files.pythonhosted.org/packages/f5/bf/3da21b6dc4c32bc8e60a4d145b1cf74944dc34557322a123795c6d13207f/awscli-1.16.34-py2.py3-none-any.whl (1.4MB)
    0% |▎                               | 10kB 27.1MB/s eta 0:00:01
    1% |▌                               | 20kB 2.0MB/s eta 0:00:01
    2% |▊                               | 30kB 2.3MB/s eta 0:00:01
    3% |█                               | 40kB 2.1MB/s eta 0:00:01
---
[00:46:27] failures:
[00:46:27] 
[00:46:27] ---- [ui] ui/iterators/array-of-ranges.rs stdout ----
[00:46:27] 
[00:46:27] error: Error: expected failure status (Some(1)) but received status Some(101).
[00:46:27] status: exit code: 101
[00:46:27] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/iterators/array-of-ranges.rs" "--target=i586-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/iterators/array-of-ranges/a" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/i586-unknown-linux-gnu/native/rust-test-helpers" "-Clinker=cc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/iterators/array-of-ranges/auxiliary" "-A" "unused"
[00:46:27] ------------------------------------------
[00:46:27] 
[00:46:27] ------------------------------------------
[00:46:27] stderr:
---
[00:46:27] note: the compiler unexpectedly panicked. this is a bug.
[00:46:27] 
[00:46:27] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[00:46:27] 
[00:46:27] note: rustc 1.31.0-nightly (d502144fd 2018-10-16) running on x86_64-unknown-linux-gnu
[00:46:27] 
[00:46:27] note: compiler flags: -Z ui-testing -Z unstable-options -C prefer-dynamic -C rpath -C linker=cc
[00:46:27] 
[00:46:27] ------------------------------------------
[00:46:27] 
[00:46:27] thread '[ui] ui/iterators/array-of-ranges.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3277:9
---
[00:46:27] 
[00:46:27] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:501:22
[00:46:27] 
[00:46:27] 
[00:46:27] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/i586-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-i586-unknown-linux-gnu" "--mode" "ui" "--target" "i586-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--linker" "cc" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/i586-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--llvm-version" "8.0.0svn\n" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[00:46:27] 
[00:46:27] 
[00:46:27] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --target i586-unknown-linux-gnu,i686-unknown-linux-musl
[00:46:27] Build completed unsuccessfully in 0:43:14
---
travis_time:end:02799a0c:start=1539655478254320481,finish=1539655478260572730,duration=6252249
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:3e37701a
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:2a29cbf4
travis_time:start:2a29cbf4
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:23c122f8
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

estebank commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -410,6 +409,37 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+                    Some(format!("[{}]", self.tcx.type_of(def.did).to_string())),
+                ));
+                if let Some(len) = len.val.try_to_scalar().and_then(|scalar| {
+                    scalar.to_i64().ok()
```
Hi @oli-obk!

Could you take a look at how I'm using `to_i64` here? In travis I'm getting the following test error, and looking at [the interface of `Scalar`](https://github.com/rust-lang/rust/blob/master/src/librustc/mir/interpret/value.rs#L114) I don't see a safe way of figuring wether the value is `1` without tripping the assert on some platforms. I need to check wether `len.val` is `1` (at the very least) or the actual value (more flexibility, but not needed now).

```
[00:46:27] thread 'main' panicked at 'assertion failed: `(left == right)`
[00:46:27]   left: `8`,
[00:46:27]  right: `4`', librustc/mir/interpret/value.rs:242:17
[00:46:27] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:46:27] 
[00:46:27] error: internal compiler error: unexpected panic
```

oli-obk replied:
You can use [`scalar.to_usize(self.tcx).ok()`](https://github.com/rust-lang/rust/blob/master/src/librustc/mir/interpret/value.rs#L312) which actually takes the target platform's `usize` size and should thus work on all platforms

estebank replied:
Thanks! I didn't realize that `TyCtxt` implemented `HasDataLayout`.

estebank commented:
@bors r-

estebank commented:
@bors r=varkor

bors commented:
:pushpin: Commit def0f544a31bc6cc3dbf153f28304219fa38fa5c has been approved by `varkor`

<!-- @bors r=varkor def0f544a31bc6cc3dbf153f28304219fa38fa5c -->

bors commented:
:hourglass: Testing commit def0f544a31bc6cc3dbf153f28304219fa38fa5c with merge 1dceaddfbe163e2d916c904b98923342730ba970...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19576451), [status-travis](https://travis-ci.org/rust-lang/rust/builds/442640740?utm_source=github_status&utm_medium=notification)
Approved by: varkor
Pushing 1dceaddfbe163e2d916c904b98923342730ba970 to master...

clarcharr commented:
Late but... I think this just replaced several clippy lints with just one diagnostic change.
